language/move-prover/bytecode/src/loop_analysis.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    function_data_builder::FunctionDataBuilder,
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    graph::{Graph, Reducible},
+    stackless_bytecode::{
+        BorrowNode,
+        Bytecode::{self, *},
+        Label, Operation, PropKind,
+    },
+    stackless_control_flow_graph::{BlockContent, BlockId, StacklessControlFlowGraph},
+};
+use move_model::{ast, model::FunctionEnv, ty::BOOL_TYPE};
+use std::collections::{BTreeMap, BTreeSet};
+
+#[derive(Debug, Default)]
+pub struct LoopAnnotation {
+    pub loop_targets: BTreeMap<Label, BTreeSet<usize>>,
+}
+
+pub struct LoopAnalysisProcessor {}
+
+impl LoopAnalysisProcessor {
+    pub fn new() -> Box<Self> {
+        Box::new(LoopAnalysisProcessor {})
+    }
+}
+
+impl FunctionTargetProcessor for LoopAnalysisProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv<'_>,
+        data: FunctionData,
+    ) -> FunctionData {
+        if func_env.is_native() {
+            return data;
+        }
+        let loop_targets = Self::collect_loop_targets(func_env, &data);
+        let mut data = Self::transform(func_env, data, &loop_targets);
+        let loop_annotation = LoopAnnotation { loop_targets };
+        data.annotations.set::<LoopAnnotation>(loop_annotation);
+        data
+    }
+
+    fn name(&self) -> String {
+        "loop_analysis".to_string()
+    }
+}
+
+impl LoopAnalysisProcessor {
+    fn transform(
+        func_env: &FunctionEnv<'_>,
+        mut data: FunctionData,
+        loop_targets: &BTreeMap<Label, BTreeSet<usize>>,
+    ) -> FunctionData {
+        let code = std::mem::take(&mut data.code);
+        let mut builder = FunctionDataBuilder::new(func_env, data);
+        for bytecode in code {
+            match bytecode {
+                Label(attr_id, label) => {
+                    builder.emit(bytecode);
+                    builder.set_loc_from_attr(attr_id);
+                    if loop_targets.contains_key(&label) {
+                        let attr_id = builder.new_attr();
+                        builder.emit(Bytecode::Prop(
+                            attr_id,
+                            PropKind::Assume,
+                            builder.mk_call(
```

Comment:
I think there is `builder.mk_bool_call` which saves you providing the type. For this example, I guess there is even `builder.mk_not`. The goal is essentially that most expressions we build are constructed via function in FunctionDataBuilder. This way we could implement simplifications there.